networkx/utils/decorators.py

```diff
@@ -450,22 +361,536 @@ def random_array(dims, random_state=1):
+    transform an argument and return a closing function:
+
+        def open_file(fn):
+            f = open(fn)
+            return f, lambda: f.close()
+
+        @argmap.try_finally(open_file, 'file')
+        def foo(file):
+            print(file.read())
+
+    is equivalent to
+
+        def foo(file):
+            file, close_file = open_file(file)
+            try:
+                print(file.read())
+            finally:
+                close_file()
+
+    """
+
+    def __init__(self, func, *args):
+        self._func = func
+        self._args = args
+        self._finally = False
+
+    @classmethod
+    def try_finally(cls, func, *args):
+        """Alternative decorator-constructor which first transforms one or more
+        arguments with func, executes the function to be decorated in a try
+        block, and then calls a cleanup function in the associated finally
+        block.  The cleanup function takes no arguments, and is expected to be
+        the second value returned by func.
+
+        Parameters
+        ----------
+        func : callable
+            The function to apply to arguments
+
+        *params: iterable of (int, str or tuple)
+            A list of parameters, specified either as strings (their names),
+            ints (numerical indices) or arbitrarily nested tuples thereof.
+
+        Examples
+        --------
+
+            def open_close(filename):
+                file = open(filename)
+                return file, lambda: file.close()
+
+            @argmap.try_finally(open_close, 0)
+            def print_file(file):
+                print(file.read())
+
+        generates a decorated function
+
+            def print_file(filename):
+                file, close = open_close(filename)
+                try:
+                    print(file.read())
+                finally:
+                    close()
+
+        """
+        dec = cls(func, *args)
+        dec._finally = True
+        return dec
+
+    @staticmethod
+    def _lazy_compile(func):
+        """Assemble and compile the source of our optimized decorator, and
+        intrusively replace its code with the compiled version's.
+
+        Parameters
+        ----------
+        func : callable
+            A function returned by argmap.__call__ which is in the process
+            of being called for the first time.
+
+        Returns
+        -------
+        func : callable
+            The same function, with a new __code__ object.
+
+        """
+        real_func = func.__argmap__.compile(func.__wrapped__)
+        func.__code__ = real_func.__code__
+        func.__globals__.update(real_func.__globals__)
+        func.__dict__.update(real_func.__dict__)
+        return func
+
+    def __call__(self, f):
+        """Construct a lazily decorated wrapper of f.  The decorated function
+        will be compiled when it is called for the first time, and it will
+        replace its own __code__ object so subsequent calls are fast.
```

Comment:
```suggestion
        """Construct a lazily decorated wrapper of f.  
        
        The decorated function will be compiled when it is called for the first time,
        and it will replace its own __code__ object so subsequent calls are fast.
```